src/ILCompiler.Compiler/src/Compiler/BlockedInternalsBlockingPolicy.cs

```diff
@@ -140,21 +149,24 @@ private bool ComputeIsBlocked(EcmaType type, ModuleBlockingMode blockingMode)
         }
         private BlockedTypeHashtable _blockedTypes;
 
-        private MetadataType _arrayOfTType;
-        private MetadataType InitializeArrayOfTType(TypeSystemEntity contextEntity)
-        {
-            _arrayOfTType = contextEntity.Context.SystemModule.GetType("System", "Array`1");
-            return _arrayOfTType;
-        }
-        private MetadataType GetArrayOfTType(TypeSystemEntity contextEntity)
+        private MetadataType[] _knownTypes;
+
+        private MetadataType InitializeKnownTypes(TypeSystemEntity entity, int index)
         {
-            if (_arrayOfTType != null)
+            _knownTypes = new MetadataType[]
             {
-                return _arrayOfTType;
-            }
-            return InitializeArrayOfTType(contextEntity);
+                entity.Context.SystemModule.GetType("System", "Array`1"),
+                entity.Context.SystemModule.GetType("System.Runtime.Serialization", "SerializationInfo", false),
+                entity.Context.SystemModule.GetType("System.Runtime.Serialization", "ISerializable", false),
+            };
+
+            return _knownTypes[index];
         }
 
+        private MetadataType GetArrayOfTType(TypeSystemEntity contextEntity) => _knownTypes?[0] ?? InitializeKnownTypes(contextEntity, 0);
```

Comment:
Should the constructor rather take `TypeSystemContext` to avoid this lazy initialization dances?